src/collapse/dtrace.rs

```diff
@@ -109,21 +151,152 @@ impl Collapse for Folder {
+    fn set_nstacks_per_job(&mut self, n: usize) {
+        self.opt.nstacks_per_job = n;
+    }
+
+    #[cfg(test)]
+    fn set_nthreads(&mut self, n: usize) {
+        self.opt.nthreads = n;
     }
 }
 
 impl Folder {
+    fn collapse_single_threaded<R>(&mut self, mut reader: R) -> io::Result<()>
+    where
+        R: io::BufRead,
+    {
+        let mut line = String::new();
+        loop {
+            line.clear();
+
+            if reader.read_line(&mut line)? == 0 {
+                break;
+            }
+
+            let line = line.trim();
+
+            if line.is_empty() {
+                continue;
+            } else if let Ok(count) = line.parse::<usize>() {
+                self.on_stack_end(count);
+            } else {
+                self.on_stack_line(line);
+            }
+        }
+        Ok(())
+    }
+
+    fn collapse_multi_threaded<R>(&mut self, mut reader: R) -> io::Result<()>
+    where
+        R: io::BufRead,
+    {
+        assert!(self.occurrences.is_concurrent());
+        assert!(self.opt.nstacks_per_job > 0);
+        assert!(self.opt.nthreads > 1);
+
+        enum Message {
+            Job(Vec<u8>),
+            Shutdown,
+        }
+
+        crossbeam::thread::scope(|scope| {
+            // Spin up the threadpool / worker threads.
+            let (tx_input, rx_input) = channel::unbounded();
+            let (tx_output, rx_output) = channel::unbounded();
+            let mut handles = Vec::with_capacity(self.opt.nthreads);
+            for _ in 0..self.opt.nthreads {
+                let rx_input = rx_input.clone();
+                let tx_output = tx_output.clone();
+
+                let occurrences = self.occurrences.clone();
+                let opt = self.opt.clone();
+
+                let handle = scope.spawn(move |_| {
+                    let mut folder = Folder {
+                        cache_inlines: Vec::default(),
+                        occurrences,
+                        stack: VecDeque::default(),
+                        stack_str_size: 0,
+                        opt,
+                    };
+                    loop {
+                        // Receive the data...
+                        let data = match rx_input.recv().unwrap() {
+                            Message::Job(data) => data,
+                            Message::Shutdown => return,
+                        };
+                        // Process the data...
+                        let result = folder.collapse_single_threaded(&data[..]);
+                        // Send the result...
+                        tx_output.send(result).unwrap();
+                        // Prepare for the next round...
+                        folder.stack.clear();
+                        folder.stack_str_size = 0;
+                    }
+                });
+                handles.push(handle);
+            }
+
+            // State for the loop...
+            let buf_capacity = collapse::smallest_power_of_two_larger_than(
```

Comment:
Isn't this the same as [`usize::next_power_of_two`](https://doc.rust-lang.org/std/primitive.usize.html#method.next_power_of_two)?

Reply:
Cool. I didn't know about that! Will obviously use the built-in instead.